packages/common/core-interfaces/src/handles.ts

```diff
@@ -117,6 +118,52 @@ export interface IFluidHandleInternalPayloadPending<
 	readonly payloadPending: boolean;
 }
 
+/**
+ * The state of the handle's payload.
+ * - "local" - The payload is only available to the local client, and not to remote collaborators
+ * - "shared" - The payload is availabe to both the local client and remote collaborators
+ * - "pending" - The payload is not yet available to the local client
+ * - "failed" - The payload is available to the local client but has failed in sharing to remote collaborators
+ * @legacy
+ * @alpha
+ */
+export type PayloadState = "local" | "shared" | "pending" | "failed";
+
+/**
+ * Events which fire as the handle's payload state transitions.
+ * @legacy
+ * @alpha
+ */
+export interface IFluidHandlePayloadPendingEvents extends IEvent {
+	/**
+	 * Emitted when the payload becomes available to all clients.
+	 */
+	(event: "shared", listener: () => void);
+	/**
+	 * Emitted for locally created handles when the payload fails sharing to remote collaborators.
+	 */
+	(event: "failed", listener: (error: unknown) => void);
```

Comment:
i think we should remove both of these, and just have a single event for progress. Maybe something like:
`(event: "progress", listener: ({type:"update"} | {type:"complete"} | {type:"failed" error?:Error})=>void)`

this ensure the event is applicable to all clients, and it lets the handle implementation short-circuit the event if someone registers while the state is already terminal (complete or failed). we do this for a number of events like connected, as we saw users frequently building dangling event handles that never resolved as the connection state was already connected.

something like progress is probably optional for now, but someday i could see us using signals in the blob manager so broadcast progress updates.

Reply:
I hesitate to merge since the customer reaction to a failure likely has no overlap with their reaction to successful completion.  In general it's more ergonomic for a customer to register specifically for the event they're interested in observing (rather than forcing them to conditionalize within their handler).

The short-circuiting behavior you mention is an antipattern, we removed many of those over time (e.g. #14371, not sure if we got them all though).  See https://github.com/microsoft/FluidFramework/pull/14272#issuecomment-1440443768 and other comments in that PR for more context.

Reply:
i find it more ergonomic to only need a single event when both events must always be considered, as i don't think there is ever a case where shared or failed are useful separately, so the current design forces a complicated multi event pattern. if there is a single event pattern, combining also doesn't make that more difficult.